File: data_structures/binary_search_tree.java
// A Binary Search Tree is a binary tree data structure where each node has at most two children (left and right), and for each node:
// - The left child's key is less than the node's key.
// - The right child's key is greater than the node's key.
// This property enables efficient search, insert, and delete operations.

// Think of a phonebook where entires are stored alphabetically by last name.
// Each entry (node) contains a last name (key) and associated information (value).
// Searching for a specific last name efficiently is analogous to using a BST.

class TreeNode {
    int key;
    TreeNode left;
    TreeNode right;

    public TreeNode(int key) {
        this.key = key;
        this.left = null;
        this.right = null;
    }
}

public class BinarySearchTree {
    private TreeNode root;

    public BinarySearchTree() {
        this.root = null;
    }

    public void insert(int key) {
        root = insertRec(root, key);
    }

    private TreeNode insertRec(TreeNode root, int key) {
        if (root == null) {
            root = new TreeNode(key);
            return root;
        }

        if (key < root.key) {
            root.left = insertRec(root.left, key);
        } else if (key > root.key) {
            root.right = insertRec(root.right, key);
        }

        return root;
    }

    public boolean search(int key) {
        return searchRec(root, key);
    }

    public boolean searchRec(TreeNode root, int key) {
        if (root == null) {
            return false;
        }

        if (root.key == key) {
            return true;
        } else if (key < root.key) {
            return searchRec(root.left, key);
        } else {
            return searchRec(root.right, key);
        }
    }

    public static void main(String[] args) {
        BinarySearchTree bst = new BinarySearchTree();

        bst.insert(50);
        bst.insert(30);
        bst.insert(70);
        bst.insert(20);
        bst.insert(40);

        System.out.println("Search for 30: " + bst.search(30));
        System.out.println("Search for 60: " + bst.search(60));
    }
}

File: data_structures/tree/binary_search_tree.kt
// A Binary Search Tree is a Binary Tree with an ordering property: left child < parent < right child

class BinarySearchTreeNode(val value: Int) {
    var left: BinarySearchTreeNode? = null
    var right: BinarySearchTreeNode? = null
}

class BinarySearchTree {
    var root: BinarySearchTreeNode? = null

    // Insert a new node into the BST
    fun insert(value: Int) {
        root = insertRec(root, value)
    }

    private fun insertRec(node: BinarySearchTreeNode?, value: Int): BinarySearchTreeNode {
        if (node == null) return BinarySearchTreeNode(value)

        if (value < node.value) node.left = insertRec(node.left, value)
        else if (value > node.value) node.right = insertRec(node.right, value)
        return node
    }

    fun search(value: Int): Boolean {
        return searchRec(root, value) != null
    }

    private fun searchRec(node: BinarySearchTreeNode?, value: Int): BinarySearchTreeNode {
        if (node == null || node.value == value) return node

        return if (value < node.value) searchRec(node.left, value)
        else searchRec(node.right, value)
    }

    // In-order traversal: Left -> Node -> Right
    fun inOrder(node: BinarySearchTreeNode?) {
        if (node != null) {
            inOrder(node.left)
            print("${node.value} ")
            inOrder(node.right)
        }
    }
}

fun main() {
    val bst = BinarySearchTree()
    bst.insert(10)
    bst.insert(5)
    bst.insert(15)
    bst.insert(2)
    bst.insert(7)

    println("In-order Traversal")
    bst.inOrder(bst.root) // Output: 2 5 7 10 15
    println("\nSearch for 7: ${bst.search(7)}") // Output: true
    println("Search for 12: ${bst.search(12)}") // Output: false
}
